### ISSUE TITLE
[Technical Question] Difference between subscribeToDesiredProperties and subscribeToTwinDesiredProperties on DeviceClient

### ISSUE THREAD

**Issue by jokarl** [Azure/azure-iot-sdk-java#916]
I'm working on implementing twin property callbacks for our devices.

I want to subscribe to a property in a twin that is an object:

```
"config": {

}
```

However, it is not clear from examples or documentation what the actual difference is between these two subscription methods. To my eyes, it looks like `subscribeToDesiredProperties` is a better fit and cleaner to use, because I can say what type of object the config property is:

```java
final var desiredProperties = new HashMap<Property, Pair<PropertyCallBack<String, CustomerConfiguration>, Object>>();
desiredProperties.put(new Property("config", null), new Pair<>((propertyKey, propertyValue, context) -> {
    // Known value type, nice
    System.out.println(propertyValue instanceof CustomerConfiguration);
}, null));

deviceClient.subscribeToDesiredProperties(desiredProperties);
```

Meanwhile, the `TwinPropertyCallBack` isn't parameterized and  neither is the `Property` that is received, so what do I receive when a property is updated in the twin?
```java
final  var twinProperties = new HashMap<Property, Pair<TwinPropertyCallBack, Object>>();
twinProperties.put(new Property("config", null), new Pair<>((property, context) -> {
   // What even is the type of property.getValue()?
   // Is it an object node? Is it some Azure specific type?
}, null));

deviceClient.subscribeToTwinDesiredProperties(twinProperties);
```

If you could shine a light on this and tell me the right way to listen to device twin changes, I'd be grateful.

**Bottom line**
What is the difference between `subscribeToDesiredProperties` and `subscribeToTwinDesiredProperties` on `DeviceClient`?

AB#8306036

**Comment by timtay-microsoft** [comment]
Both of the subscription methods should do the same thing within the SDK, but they just have differently shaped callbacks. Both methods will ensure that you get called back when the desired property you set a callback for is updated.

The Property class is shared between reported and desired properties, so that may be causing some confusion here. The value of a Property doesn't need to be assigned when you are just subscribing to desired properties. After all, you aren't expected to know the initial value. You only know about the key.

```java
Map<Property, Pair<TwinPropertyCallBack, Object>> desiredProperties = new HashMap<Property, Pair<TwinPropertyCallBack, Object>>()
{
    {
        put(new Property("HomeTemp(F)", null), new Pair<TwinPropertyCallBack, Object>(new onHomeTempChange(), null));
        put(new Property("LivingRoomLights", null), new Pair<TwinPropertyCallBack, Object>(new onProperty(), null));
        put(new Property("BedroomRoomLights", null), new Pair<TwinPropertyCallBack, Object>(new onProperty(), null));
        put(new Property("HomeSecurityCamera", null), new Pair<TwinPropertyCallBack, Object>(new onCameraActivity(), null));
    }
};
client.subscribeToTwinDesiredProperties(desiredProperties);
```

It only needs to be assigned when you are sending reported properties
```java
Set<Property> reportProperties = new HashSet<Property>()
{
    {
        add(new Property("HomeTemp(F)", 70));
        add(new Property("LivingRoomLights", LIGHTS.ON));
        add(new Property("BedroomRoomLights", LIGHTS.OFF));
    }
};
client.sendReportedProperties(reportProperties);
```

When you receive a desired property update through the TwinPropertyCallback, the value of the property could be a few different types. It may just be a String if your property is a simple "someKey": "someValue", or it could be an int or boolean. It may also be a map if the value is a nested json object, as it will be for your config object. 

```java
protected static class onProperty implements TwinPropertyCallBack
{
    @Override
    public void TwinPropertyCallBack(Property property, Object context)
    {
        System.out.println(
                "onProperty callback for " + (property.getIsReported()?"reported": "desired") +
                        " property " + property.getKey() +
                        " to " + property.getValue() +
                        ", Properties version:" + property.getVersion());
    }
}
```

Historically, the SDK only had ```subscribeToDesiredProperties```, but there were some limitations to it, so later on ```subscribeToTwinDesiredProperties``` was added. In general, the newer API is probably the one you want to use, but they both certainly work.

Does that help to clear it up a bit?

**Comment by jokarl** [comment]
Thank you for the thorough explanation. I ended up using the `TwinPropertyCallBack` interface. 

I experimented a bit and found documentation and examples a bit lacking in some areas. For instance I didn't find any documentation about the data received in the value of the property when using the `TwinPropertyCallBack` interface. 

After some experimenting I saw that single values (i.e. not objects) in the twin are actually primitives when getting the value of the property. Objects on the other hand are of type `TwinCollection` and need to be serialized and then deserialized to get an object from.

**Comment by timtay-microsoft** [comment]
We'll try to spruce up our samples a bit so that these APIs are less confusing then
